### DISCUSSION TITLE
Confused about behavior of CIECAM at low L_A and Y_b

### DISCUSSION THREAD

**Discussion by mesvam** [colour-science/colour#1226] [Q&A] [answered]
At low L_A and Y_b, the model seems to break down. For example, these are the CAM coordinates with default L_A, Y_b:
```
import colour
import pandas as pd
sRGB = [
    (0, 0, 0), # black
    (0, 1, 0), # green
    (1, 1, 1)  # white
]
cam = colour.convert(sRGB, 'sRGB', 'CIECAM16',
    XYZ_w=colour.TVS_ILLUMINANTS["CIE 1931 2 Degree Standard Observer"]["D65"]/100,
    L_A=64*0.2/np.pi, Y_b=20,
)
pd.DataFrame(cam, columns=['J', 'C', 'h', 's', 'Q', 'M', 'H', 'HC'])
```
```
              J             C         h             s             Q             M         H  HC
0  3.258490e-24  6.221649e-26  0.500000  1.419209e-07  2.233087e-12  4.497779e-26  0.561821 NaN
1  7.910358e-01  1.081873e+00  0.395067  8.431147e-01  1.100260e+00  7.821118e-01  0.443016 NaN
2  1.000010e+00  3.082042e-02  0.583308  1.342041e-01  1.237085e+00  2.228082e-02  0.666476 NaN
```

But with very low L_A, M tends towards zero. This doesn't make sense, since color doesn't go away when viewing in a dark environment. (I can't use L_A=0 or Y_b=0 to demonstrate because of numerical errors, see bug #1227)

```
cam = colour.convert(sRGB, 'sRGB', 'CIECAM16',
    XYZ_w=colour.TVS_ILLUMINANTS["CIE 1931 2 Degree Standard Observer"]["D65"]/100,
    L_A=1e-30, Y_b=20,
)
pd.DataFrame(cam, columns=['J', 'C', 'h', 's', 'Q', 'M', 'H', 'HC'])
```
```
              J             C         h             s             Q             M         H  HC
0  8.734361e-08  1.018622e-17  0.500000  3.855345e-07  2.167138e-12  3.221165e-25  0.561821 NaN
1  7.875197e-01  3.343898e-10  0.395617  4.031120e-05  6.507311e-09  1.057433e-17  0.443577 NaN
2  1.000007e+00  1.354441e-11  0.583176  7.642647e-06  7.332842e-09  4.283117e-19  0.666321 NaN
```

Similarly, for low Y_b; C, Q, and M goes to infinity. Which is equally perplexing, since a spot of color on a black background isn't really perceived as an explosion of color.

```
cam = colour.convert(sRGB, 'sRGB', 'CIECAM16',
    XYZ_w=colour.TVS_ILLUMINANTS["CIE 1931 2 Degree Standard Observer"]["D65"]/100,
    L_A=64*0.2/np.pi, Y_b=1e-30,
)
pd.DataFrame(cam, columns=['J', 'C', 'h', 's', 'Q', 'M', 'H', 'HC'])
```
```
              J             C         h             s             Q             M         H  HC
0  9.187890e-19  1.147554e-17  0.500000  6.667251e-08  1.866260e-03  8.295944e-18  0.561821 NaN
1  8.352566e-01  3.861502e+05  0.395067  3.960840e-01  1.779403e+06  2.791572e+05  0.443016 NaN
2  1.000007e+00  1.070548e+04  0.583308  6.304728e-02  1.947000e+06  7.739247e+03  0.666476 NaN
```

Am I misinterpreting the meaning of adapting field luminance L_A and background luminance Y_b?

**Comment by mesvam** [comment]
@tjdcs posted a good answer here: https://github.com/colour-science/colour/issues/1227#issuecomment-1832554764

**Comment by mesvam** [comment]
I see from https://cielab.xyz/pdf/CIECAM02_and_Its_Recent_Developments.pdf that CIECAM02 is designed for related colors (e.g. an image) and is not valid for unrelated colors (lone spot of color on a dark background). 

However, there is a modification by [Fu et al (2007) "Quantifying colour appearance for unrelated colour under photopic and mesopic vision"](https://library.imaging.org/admin/apis/public/api/ist/website/downloadArticle/cic/15/1/art00060) to make it work in that situation, but as far as I can tell, M still goes to 0 in the limit of total darkness since the only modification to M is a multiplicative factor. So it does not really explain it either.

**Comment by tjdcs** [comment]
There is also CAM16u from Ronnier Luo’s research group(s)

But tread lightly and believe what you see in your environment. These
models are not nearly as well tested and refined.

**Comment by mesvam** [comment, marked as answer]
After reading a handful of related literature, it seems that L_A is often taken as average luminance of the image (which is often around 20% of reference white), rather than the luminance of the whole field of vision excluding the stimulus (proximal field + background + surround), as specified

I guess this makes sense from the perspective of the model, since L_A is supposed to represent the brightness level that the viewer has adapted to. If you're staring at a large image, a darker surrounding doesn't have much of an effect on adaptation. With that interpretation, setting L_A = 0 means the stimulus would have to be completely dark as well (if there is any light at all, the eyes will adapt to an L_A ≠ 0), so it makes sense that since you don't see anything at all in total darkness, there is no perception of color.

I don't have a good explanation for the behavior of Y_b = 0 yet

**Reply by tjdcs** [reply]
Y_b tends the Lw or adaptation white point towards infinity. Which also doesn't make any sense.

**Reply by mesvam** [reply]
If Y_b = 100 L_b / L_w, then Y_b -> 0 only forces L_b -> 0, but L_w can take any finite value

**Reply by tjdcs** [reply]
Lw = L_a  * 100 / Y_b IIRC. But I might be remembering something else too. You are the one in the code right now. :)

**Comment by tjdcs** [comment]
Yes, and CAM16u from Ronnie Luo's research group(s)

But tread lightly. Unrelated colors in general are much less well understood and these models get much less attention. You should also read Hao Xie's dissertation research. 

Cheers!

**Reply by mesvam** [reply]
Wonderful, thank you very much for the suggestions and further reading!
